include/glow/Graph/TensorLayout.h

```diff
@@ -0,0 +1,194 @@
+
+#include <array>
+#include <memory>
+#include <string>
+
+#include "glow/Graph/Nodes.h"
+#include "glow/Support/Error.h"
+
+namespace glow {
+
+/// Layout requirements's Singleton.
+template <typename T> class TensorLayoutSingleton {
+public:
+  /// This is how the verifier, Backend and post-loading canonicalizer can
+  /// access layout constraints.
+  static T &getInstance() {
+    // The Ctor will only be called once.
+    static const std::unique_ptr<T> instance{new T{token_{}}};
+    return *instance;
+  }
+
+protected:
+  /// Allow the base class to call any subclass's constructor.
+  struct token_ {};
+
+  /// Default Ctor.
+  TensorLayoutSingleton() {}
+
+  /// Dtor.
+  virtual ~TensorLayoutSingleton() {}
+
+private:
+  /// Delete copy constructor.
+  TensorLayoutSingleton(const TensorLayoutSingleton &) = delete;
+
+  /// Delete move constructor.
+  TensorLayoutSingleton(TensorLayoutSingleton &&) = delete;
+
+  /// Delete copy assignment.
+  TensorLayoutSingleton &operator=(const TensorLayoutSingleton &) = delete;
+
+  /// Delete move assignment.
+  TensorLayoutSingleton &operator=(TensorLayoutSingleton &&) = delete;
+};
+
+/// TensorLayoutDescription - optional helper class for parsing string-based
+/// layout.
+class TensorLayoutDescription {
+  /// Tensor dimensions descriptions for all dimensions.
+  std::string dims_[max_tensor_dimensions];
+  /// The serialization of the layout.
+  std::string serializedLayout_;
+  /// Expected number of dimensions.
+  size_t numDims_;
+
+public:
+  virtual ~TensorLayoutDescription() = default;
+  /// Constructs this helper class from a serialized string representation.
+  TensorLayoutDescription(std::string text);
+  /// Constructs this helper class from an array of strings representing each
+  /// individual / pre-separated dimension.
+  TensorLayoutDescription(llvm::ArrayRef<std::string> dims);
+  /// \returns the alignment of a dimension \p n.
+  size_t getAlignment(size_t n) const;
+  /// \returns the alignment by parsing dimension string \p s.
+  size_t getAlignment(const std::string &s) const;
+  /// \returns true if both tensor layouts are the same.
+  bool isSameLayout(const TensorLayoutDescription &rhs) const;
+  /// \returns description of the dimension \p n.
+  const llvm::StringRef getNthDimDescription(size_t n) const;
+  /// \returns the description of all dimensions.
+  llvm::ArrayRef<std::string> getDims() const;
+  /// \returns number of dimensions.
+  size_t getNumDims() const { return numDims_; }
+  /// \returns layout name.
+  llvm::StringRef getSerializedLayout() const { return serializedLayout_; }
+  /// \returns true if the layout is "*" in all dimensions.
+  bool isAnyLayout();
+  std::string getDebugDesc() const;
+
+protected:
+  /// parse helper: get the custom extensions information. the default, virtual,
+  /// implementation just ignores all the data until the end token.
+  virtual void parseCustomExtensions(llvm::StringRef &text, unsigned idx);
```

Comment:
Seeing `llvm::StringRef &` is a bit unusual. But I see why I'd use it. Essentially it represents your parsing context. Hopefully it is enough for the future use-cases, otherwise we'd need to extend these APIs.